app/src/main/java/com/duckduckgo/app/global/DefaultRoleBrowserDialogExperiment.kt

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright (c) 2020 DuckDuckGo
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.app.global
+
+import android.app.role.RoleManager
+import android.content.Context
+import android.content.Intent
+import android.os.Build
+import com.duckduckgo.app.global.install.AppInstallStore
+import com.duckduckgo.app.statistics.VariantManager
+import timber.log.Timber
+
+interface DefaultRoleBrowserDialogExperiment {
+    fun createIntent(context: Context): Intent?
+    fun shouldShowExperiment(): Boolean
+    fun experimentShown()
+}
+
+class RealDefaultRoleBrowserDialogExperiment(
+    private val appInstallStore: AppInstallStore,
+    private val variantManager: VariantManager
+) : DefaultRoleBrowserDialogExperiment {
+
+    /**
+     * @return an Intent to launch the role browser dialog
+     */
+    override fun createIntent(context: Context): Intent? {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+            val roleManager = context.getSystemService(RoleManager::class.java) ?: return null
+
+            val isRoleAvailable = roleManager.isRoleAvailable(RoleManager.ROLE_BROWSER)
+            if (isRoleAvailable) {
+                val isRoleHeld = roleManager.isRoleHeld(RoleManager.ROLE_BROWSER)
+                return if (!isRoleHeld) {
+                    roleManager.createRequestRoleIntent(RoleManager.ROLE_BROWSER)
+                } else {
+                    Timber.i("Browser role held")
+                    null
+                }
+            }
+        }
+        return null
+    }
+
+    override fun shouldShowExperiment(): Boolean {
```

Comment:
question: is there any way to reproduce this in our app?

Reply:
Not right now, because we are adding the new (role manager) dialog just to onboarding. So the user will only see it once. It will never happen that count is above `1`
I placed this as a safeguard in case we want to use it also in settings in the near future--which was discussed in the near future.

The comment in that method gives the background as to why that safeguard is there.